api/README.md

```diff
@@ -0,0 +1,69 @@
+# API Development
+
+This documentation is for developers to extend the current APIs. For the
+administrative documentation on the APIs, see [SETUP/API.md](SETUP/API.md).
+
+## Design
+
+The DP API uses a RESTful design, where URLs uniquely access resources.
+
+### Routing work to functions
+
+The framework is designed for a single function to handle all actions for
+a specific portion of the URI. If there are additional pieces to the URI,
+the `api_router()` is called to handle the remainder. This is achieved
+by using well-defined function names that reflect the URI being handled.
+
+Take the following URI:
+```
+/api/v1/projects/<projectid>/wordlists/<good>
+```
+
+After handling authentication and rate limiting, the router dispatches
+the initial work to `api_v1()`. That function does not consume part of the
+URI or query parameters so it calls back to the router. The router then
+dispatches the remaining work to `api_v1_projects()`.
+
+`api_v1_projects()` checks the URI to see if there are additional parts on
+the URI. If there are, it loads the `projectid` off the URI, validates it
+and passes the Project object back to the router for the next handler.
+
+The router then calls `api_v1_projects_PROJECTID_wordlists()` which takes the
+project, determines which words list to use (`good`) and based on the method
+type either returns the word list (GET) or updates the word list (PUT). In
+both cases it returns a data structure that will be output to the user in
+JSON format.
+
+### URI-handling function responsibilities
+
+URI-handling functions are each responsible for managing the following
+items:
+
+1. Pulling any resource variables off the URI and validating them.
+2. If the URI still contains other elements, calling `api_router()` to handle
+   them, passing along resources variables or objects created from them.
+3. Validating that the requested method is supported by calling
+   `api_validate_method()` with the supported methods.
```

Comment:
Insert permission check here (later)?

Reply:
This list should absolutely call out the authorization bits - I will get that updated.